c/extras/ftrl.cc

```diff
@@ -0,0 +1,527 @@
+//------------------------------------------------------------------------------
+// Copyright 2018 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#define dt_EXTRAS_FTRL_cc
+#include "extras/ftrl.h"
+#include "frame/py_frame.h"
+#include "py_utils.h"
+#include "python/obj.h"
+#include "rowindex.h"
+#include "types.h"
+#include "utils/parallel.h"
+#include "utils/shared_mutex.h"
+#include <stdio.h>
+
+
+/*
+*  Read data from Python, train FTRL model and make predictions.
+*/
+PyObject* ftrl(PyObject*, PyObject* args) {
+
+  double a, b, l1, l2;
+  uint64_t d;
+  size_t n_epochs;
+  unsigned int seed, hash_type;
+  bool inter;
+  PyObject* arg1;
+  PyObject* arg2;
+
+  if (!PyArg_ParseTuple(args, "OOddddLnpII:ftrl", &arg1, &arg2,
+                        &a, &b, &l1, &l2, &d, &n_epochs, &inter, &hash_type,
+                        &seed)) return nullptr;
+
+  DataTable* dt_train = py::obj(arg1).to_frame();
+  DataTable* dt_test = py::obj(arg2).to_frame();
+
+  Ftrl ft(a, b, l1, l2, d, n_epochs, inter, hash_type, seed);
+  ft.train(dt_train);
+  DataTable* dt_target = ft.test(dt_test).release();
+  py::Frame* frame_target = py::Frame::from_datatable(dt_target);
+
+  return frame_target;
+}
+
+
+/*
+*  Set up FTRL parameters and initialize weights.
+*/
+Ftrl::Ftrl(double a_in, double b_in, double l1_in, double l2_in,
+           uint64_t d_in, size_t nepochs_in, bool inter_in,
+           unsigned int hash_type_in, unsigned int seed_in) :
+  a(a_in),
+  b(b_in),
+  l1(l1_in),
+  l2(l2_in),
+  d(d_in),
+  n_epochs(nepochs_in),
+  hash_type(hash_type_in),
+  seed(seed_in),
+  inter(inter_in)
+{
+  n = DoublePtr(new double[d]);
+  w = DoublePtr(new double[d]);
+  std::memset(n.get(), 0, d * sizeof(double));
+  std::memset(w.get(), 0, d * sizeof(double));
```

Comment:
instead of memset, you could have called `new double[d]()`, note the parentheses.
See https://stackoverflow.com/questions/620137

Reply:
Yes, sure. This part will probably have to be changed anyway as we may want to load these coefficients from a pre-trained model.